Data/Recursive/Propagator/Seminaive.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE MultiParamTypeClasses, TypeFamilies, FlexibleContexts #-}
+module Data.Recursive.Propagator.Seminaive
+where
+
+import qualified Data.Set as S
+import Data.Set (Set)
+import Control.Monad (unless)
+import Control.Concurrent.MVar
+import Data.IORef
+-- import Data.Maybe
+
+import Data.POrder
+
+class (Eq a, Bottom a) => ChangeAction da a where
+  -- laws:
+  -- 1. update (kickoff x) bottom == x
+  -- 2. (update dx x == y) implies ((x == y) == noChange dx x y)
+  update :: da -> a -> a
+  kickoff :: a -> da
+  noChange :: da -> a -> a -> Bool
+  noChange dx x y = x == y
+class ChangeAction (Delta a) a => Change a where type Delta a
```

Comment:
I’d also expect (maybe as a later refinement) a way to efficiently assembles deltas. I.e. a Subgroup or Monoid constraint on `da`, and a law
```
update dy (update dx x) == update (dy <> dx) x
```